### PR TITLE
[WIP] Ssl redirect

### PR DESCRIPTION
**Pull request by asridharan**
PR to add `ssl-redirect` annotation to AG ingress resource.

### CODE REVIEW

**Comment by asridharan** [comment]
This PR still needs some work. SSL redirection with basic routing works. Still need to test and work path based routing.

File: pkg/appgw/requestroutingrules.go [hunk 1]
```diff
@@ -133,18 +135,33 @@ func (builder *appGwConfigBuilder) RequestRoutingRules(ingressList [](*v1beta1.I
...
 				// need to eliminate non-unique paths
 				urlPathMaps[listenerHTTPID] = builder.pathMaps(ingress, &rule,
 					listenerHTTPID, urlPathMaps[listenerHTTPID],
 					defaultAddressPoolID, defaultHTTPSettingsID)
+
+				httpURLPathMap := urlPathMaps[listenerHTTPID]
+
+				// if ssl-redirect annotation is present setup the correct redirect configuration.
+				if annotations.IsSslRedirect(ingress) {
+					sslRedirectConfigID := builder.appGwIdentifier.redirectConfigurationID(generateSSLRedirectConfigurationName(ingress.Namespace, ingress.Name))
```
Comment (draychev) [inline pkg/appgw/requestroutingrules.go]: I would have carved this block out into a separate function to make it easier to test.

File: pkg/annotations/ingress_annotations.go [hunk 2]
```diff
@@ -35,3 +41,14 @@ func IsApplicationGatewayIngress(ing *v1beta1.Ingress) bool {
 	controllerName := ing.Annotations[IngressClassKey]
 	return controllerName == ApplicationGatewayIngressClass
 }
+
+// IsSslRedirect for HTTP end points.
+func IsSslRedirect(ing *v1beta1.Ingress) bool {
+	val, ok := ing.Annotations[SslRedirectKey]
+
+	if !ok || val != "true" {
```
[the remark this answers is not in the document]
Comment (vramakrishnan) [reply pkg/annotations/ingress_annotations.go]: There is also "comma ok " idiom, which is pretty commonly used ( https://golang.org/doc/effective_go.html#maps)

if val, ok := ing.Annotations[SslRedirectKey]; ok {
   return val  
}

File: pkg/annotations/ingress_annotations_test.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
+package annotations
```
Comment (asridharan) [inline pkg/annotations/ingress_annotations_test.go]: We need a CMAKE target to run these tests in CI?

Comment (draychev) [reply pkg/annotations/ingress_annotations_test.go]: I'll make sure to add a target for running tests.

This is work-in-progress BTW - still working on the PR...

**Comment by draychev** [comment]
This feature was merged w/ https://github.com/Azure/application-gateway-kubernetes-ingress/pull/156

**Pull request closed without merging**